.github/workflows/gh-pages.yml

```diff
@@ -28,27 +28,39 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v3
 
+      - name: Cache Setup
+        id: cache-ocaml-setup
+        uses: actions/cache@v3
+        with:
+          path: ~/.opam
+          key: ${{ runner.os }}-ocaml-setup
+
       # sphinx-multiversion uses tags to build
       # the documentation in multiple versions.
       - name: Checkout | Fetch tags
         run: git fetch --tags origin
 
       - name: Setup | mdBook | 1/2
         uses: hecrj/setup-rust-action@master
-      
+
       - name: Setup | mdBook | 2/2
         uses: peaceiris/actions-mdbook@v1
         with:
-          mdbook-version: 'latest'
+          mdbook-version: "latest"
 
       - name: Setup | Update
         run: sudo apt-get update
 
-      - name: Setup | Doxygen
-        run: sudo apt-get install doxygen
+      - name: Setup | System
+        run: sudo apt-get install doxygen python3-sphinx opam libgmp-dev
 
-      - name: Setup | Sphinx
-        run: sudo apt-get install python3-sphinx
+      - name: Setup | OCaml
+        if: steps.cache-ocaml-setup.outputs.cache-hit != 'true'
+        run: |
+          OPAMYES=true opam init --auto-setup --disable-sandboxing --yes --bare
+          opam switch create 4.12.0 --yes
+          eval $(opam env)
+          opam install --yes ocamlfind ctypes zarith cppo
```

Comment:
I'd explicitly add `odoc` too. I assume it's  currently installed as a dependency to one of these packages.